### ISSUE TITLE
%*% operator should dispatch on S4 arguments

### ISSUE THREAD

**Issue by chuanwen** [bedatadriven/renjin#356]
```
library(glmnet)
x = matrix(rnorm(3000), ncol=3)
y = rnorm(1000)
fit = glmnet(x, y)
ypred = predict(fit, x)

ERROR: error in evaluating the argument 'x' in selecting a method for function 'as.matrix'
org.renjin.eval.EvalException: Conversion = '*'
	at org.renjin.eval.Context.evaluateCall(Context.java:416)
	at org.renjin.eval.Context.evaluate(Context.java:282)
	at org.renjin.sexp.Promise.doEval(Promise.java:92)
	at org.renjin.sexp.Promise.force(Promise.java:86)
	at org.renjin.eval.Context.evaluateSymbol(Context.java:345)
	at org.renjin.eval.Context.evaluate(Context.java:278)
	at org.renjin.eval.Context.evaluate(Context.java:264)
	at org.renjin.methods.MethodDispatch.R_dispatchGeneric(MethodDispatch.java:191)
	at org.renjin.methods.MethodDispatch.standardGeneric(MethodDispatch.java:137)
	at org.renjin.methods.Methods.standardGeneric(Methods.java:369)
	at org.renjin.primitives.R$primitive$standardGeneric.doApply(R$primitive$standardGeneric.java:79)
	at org.renjin.primitives.R$primitive$standardGeneric.apply(R$primitive$standardGeneric.java:33)
	at org.renjin.eval.Context.evaluateCall(Context.java:407)
	at org.renjin.eval.Context.evaluate(Context.java:282)
	at org.renjin.eval.Context.evaluate(Context.java:264)
	at org.renjin.eval.Context.evaluate(Context.java:198)
	at org.renjin.sexp.Closure.doApply(Closure.java:97)
	at org.renjin.eval.ClosureDispatcher.apply(ClosureDispatcher.java:80)
	at org.renjin.eval.ClosureDispatcher.applyClosure(ClosureDispatcher.java:51)
	at org.renjin.sexp.Closure.apply(Closure.java:82)
	at org.renjin.eval.Context.evaluateCall(Context.java:407)
	at org.renjin.eval.Context.evaluate(Context.java:282)
	at org.renjin.eval.Context.evaluate(Context.java:264)
	at org.renjin.primitives.special.AssignLeftFunction.assignLeft(AssignLeftFunction.java:57)
	at org.renjin.primitives.special.AssignLeftFunction.apply(AssignLeftFunction.java:41)
	at org.renjin.eval.Context.evaluateCall(Context.java:407)
	at org.renjin.eval.Context.evaluate(Context.java:282)
	at org.renjin.eval.Context.evaluate(Context.java:264)
	at org.renjin.primitives.special.BeginFunction.apply(BeginFunction.java:38)
	at org.renjin.eval.Context.evaluateCall(Context.java:407)
	at org.renjin.eval.Context.evaluate(Context.java:282)
	at org.renjin.eval.Context.evaluate(Context.java:264)
	at org.renjin.eval.Calls.applyClosure(Calls.java:47)
	at org.renjin.primitives.S3$GenericMethod.doApply(S3.java:1452)
	at org.renjin.primitives.S3$GenericMethod.applyNext(S3.java:1411)
	at org.renjin.primitives.S3.NextMethod(S3.java:91)
	at org.renjin.primitives.R$primitive$NextMethod.apply(R$primitive$NextMethod.java:42)
	at org.renjin.primitives.special.InternalFunction.apply(InternalFunction.java:45)
	at org.renjin.eval.Context.evaluateCall(Context.java:407)
	at org.renjin.eval.Context.evaluate(Context.java:282)
	at org.renjin.eval.Context.evaluate(Context.java:264)
	at org.renjin.eval.Context.evaluate(Context.java:198)
	at org.renjin.sexp.Closure.doApply(Closure.java:97)
	at org.renjin.eval.ClosureDispatcher.apply(ClosureDispatcher.java:80)
	at org.renjin.eval.ClosureDispatcher.applyClosure(ClosureDispatcher.java:51)
	at org.renjin.sexp.Closure.apply(Closure.java:82)
	at org.renjin.eval.Context.evaluateCall(Context.java:407)
	at org.renjin.eval.Context.evaluate(Context.java:282)
	at org.renjin.eval.Context.evaluate(Context.java:264)
	at org.renjin.primitives.special.BeginFunction.apply(BeginFunction.java:38)
	at org.renjin.eval.Context.evaluateCall(Context.java:407)
	at org.renjin.eval.Context.evaluate(Context.java:282)
	at org.renjin.eval.Context.evaluate(Context.java:264)
	at org.renjin.eval.Calls.applyClosure(Calls.java:47)
	at org.renjin.primitives.S3$GenericMethod.doApply(S3.java:1452)
	at org.renjin.primitives.S3$GenericMethod.apply(S3.java:1401)
	at org.renjin.primitives.S3.UseMethod(S3.java:80)
	at org.renjin.primitives.S3.UseMethod(S3.java:72)
	at org.renjin.primitives.R$primitive$UseMethod.doApply(R$primitive$UseMethod.java:77)
	at org.renjin.primitives.R$primitive$UseMethod.apply(R$primitive$UseMethod.java:31)
	at org.renjin.eval.Context.evaluateCall(Context.java:407)
	at org.renjin.eval.Context.evaluate(Context.java:282)
	at org.renjin.eval.Context.evaluate(Context.java:264)
	at org.renjin.eval.Context.evaluate(Context.java:198)
	at org.renjin.sexp.Closure.doApply(Closure.java:97)
	at org.renjin.eval.ClosureDispatcher.apply(ClosureDispatcher.java:80)
	at org.renjin.eval.ClosureDispatcher.applyClosure(ClosureDispatcher.java:51)
	at org.renjin.sexp.Closure.apply(Closure.java:82)
	at org.renjin.eval.Context.evaluateCall(Context.java:407)
	at org.renjin.eval.Context.evaluate(Context.java:282)
	at org.renjin.eval.Context.evaluate(Context.java:264)
	at org.renjin.primitives.special.AssignLeftFunction.assignLeft(AssignLeftFunction.java:57)
	at org.renjin.primitives.special.AssignLeftFunction.apply(AssignLeftFunction.java:41)
	at org.renjin.eval.Context.evaluateCall(Context.java:407)
	at org.renjin.eval.Context.evaluate(Context.java:282)
	at org.renjin.eval.Context.evaluate(Context.java:264)
	at org.renjin.eval.Context.evaluateExpressionVector(Context.java:390)
	at org.renjin.eval.Context.evaluate(Context.java:280)
	at org.renjin.eval.Context.evaluate(Context.java:264)
	at org.renjin.repl.JlineRepl.readExpression(JlineRepl.java:244)
	at org.renjin.repl.JlineRepl.run(JlineRepl.java:131)
	at org.renjin.cli.Main.startInteractive(Main.java:126)
	at org.renjin.cli.Main.run(Main.java:111)
	at org.renjin.cli.Main.main(Main.java:78)
Caused by: java.util.UnknownFormatConversionException: Conversion = '*'
	at java.util.Formatter.checkText(Formatter.java:2579)
	at java.util.Formatter.parse(Formatter.java:2555)
	at java.util.Formatter.format(Formatter.java:2501)
	at java.util.Formatter.format(Formatter.java:2455)
	at java.lang.String.format(String.java:2940)
	at org.renjin.primitives.R$primitive$$percent$times$percent.doApply(R$primitive$$percent$times$percent.java:73)
	at org.renjin.primitives.R$primitive$$percent$times$percent.apply(R$primitive$$percent$times$percent.java:31)
	at org.renjin.eval.Context.evaluateCall(Context.java:407)
	... 83 more
  at as.matrix()
  at predict.glmnet()
  at NextMethod()
  at predict.elnet()
  at predict()
Caused by:org.renjin.eval.EvalException: Conversion = '*'
	at org.renjin.eval.Context.evaluateCall(Context.java:416)
	at org.renjin.eval.Context.evaluate(Context.java:282)
	at org.renjin.sexp.Promise.doEval(Promise.java:92)
	at org.renjin.sexp.Promise.force(Promise.java:86)
	at org.renjin.eval.Context.evaluateSymbol(Context.java:345)
	at org.renjin.eval.Context.evaluate(Context.java:278)
	at org.renjin.eval.Context.evaluate(Context.java:264)
	at org.renjin.methods.MethodDispatch.R_dispatchGeneric(MethodDispatch.java:191)
	at org.renjin.methods.MethodDispatch.standardGeneric(MethodDispatch.java:137)
	at org.renjin.methods.Methods.standardGeneric(Methods.java:369)
	at org.renjin.primitives.R$primitive$standardGeneric.doApply(R$primitive$standardGeneric.java:79)
	at org.renjin.primitives.R$primitive$standardGeneric.apply(R$primitive$standardGeneric.java:33)
	at org.renjin.eval.Context.evaluateCall(Context.java:407)
	at org.renjin.eval.Context.evaluate(Context.java:282)
	at org.renjin.eval.Context.evaluate(Context.java:264)
	at org.renjin.eval.Context.evaluate(Context.java:198)
	at org.renjin.sexp.Closure.doApply(Closure.java:97)
	at org.renjin.eval.ClosureDispatcher.apply(ClosureDispatcher.java:80)
	at org.renjin.eval.ClosureDispatcher.applyClosure(ClosureDispatcher.java:51)
	at org.renjin.sexp.Closure.apply(Closure.java:82)
	at org.renjin.eval.Context.evaluateCall(Context.java:407)
	at org.renjin.eval.Context.evaluate(Context.java:282)
	at org.renjin.eval.Context.evaluate(Context.java:264)
	at org.renjin.primitives.special.AssignLeftFunction.assignLeft(AssignLeftFunction.java:57)
	at org.renjin.primitives.special.AssignLeftFunction.apply(AssignLeftFunction.java:41)
	at org.renjin.eval.Context.evaluateCall(Context.java:407)
	at org.renjin.eval.Context.evaluate(Context.java:282)
	at org.renjin.eval.Context.evaluate(Context.java:264)
	at org.renjin.primitives.special.BeginFunction.apply(BeginFunction.java:38)
	at org.renjin.eval.Context.evaluateCall(Context.java:407)
	at org.renjin.eval.Context.evaluate(Context.java:282)
	at org.renjin.eval.Context.evaluate(Context.java:264)
	at org.renjin.eval.Calls.applyClosure(Calls.java:47)
	at org.renjin.primitives.S3$GenericMethod.doApply(S3.java:1452)
	at org.renjin.primitives.S3$GenericMethod.applyNext(S3.java:1411)
	at org.renjin.primitives.S3.NextMethod(S3.java:91)
	at org.renjin.primitives.R$primitive$NextMethod.apply(R$primitive$NextMethod.java:42)
	at org.renjin.primitives.special.InternalFunction.apply(InternalFunction.java:45)
	at org.renjin.eval.Context.evaluateCall(Context.java:407)
	at org.renjin.eval.Context.evaluate(Context.java:282)
	at org.renjin.eval.Context.evaluate(Context.java:264)
	at org.renjin.eval.Context.evaluate(Context.java:198)
	at org.renjin.sexp.Closure.doApply(Closure.java:97)
	at org.renjin.eval.ClosureDispatcher.apply(ClosureDispatcher.java:80)
	at org.renjin.eval.ClosureDispatcher.applyClosure(ClosureDispatcher.java:51)
	at org.renjin.sexp.Closure.apply(Closure.java:82)
	at org.renjin.eval.Context.evaluateCall(Context.java:407)
	at org.renjin.eval.Context.evaluate(Context.java:282)
	at org.renjin.eval.Context.evaluate(Context.java:264)
	at org.renjin.primitives.special.BeginFunction.apply(BeginFunction.java:38)
	at org.renjin.eval.Context.evaluateCall(Context.java:407)
	at org.renjin.eval.Context.evaluate(Context.java:282)
	at org.renjin.eval.Context.evaluate(Context.java:264)
	at org.renjin.eval.Calls.applyClosure(Calls.java:47)
	at org.renjin.primitives.S3$GenericMethod.doApply(S3.java:1452)
	at org.renjin.primitives.S3$GenericMethod.apply(S3.java:1401)
	at org.renjin.primitives.S3.UseMethod(S3.java:80)
	at org.renjin.primitives.S3.UseMethod(S3.java:72)
	at org.renjin.primitives.R$primitive$UseMethod.doApply(R$primitive$UseMethod.java:77)
	at org.renjin.primitives.R$primitive$UseMethod.apply(R$primitive$UseMethod.java:31)
	at org.renjin.eval.Context.evaluateCall(Context.java:407)
	at org.renjin.eval.Context.evaluate(Context.java:282)
	at org.renjin.eval.Context.evaluate(Context.java:264)
	at org.renjin.eval.Context.evaluate(Context.java:198)
	at org.renjin.sexp.Closure.doApply(Closure.java:97)
	at org.renjin.eval.ClosureDispatcher.apply(ClosureDispatcher.java:80)
	at org.renjin.eval.ClosureDispatcher.applyClosure(ClosureDispatcher.java:51)
	at org.renjin.sexp.Closure.apply(Closure.java:82)
	at org.renjin.eval.Context.evaluateCall(Context.java:407)
	at org.renjin.eval.Context.evaluate(Context.java:282)
	at org.renjin.eval.Context.evaluate(Context.java:264)
	at org.renjin.primitives.special.AssignLeftFunction.assignLeft(AssignLeftFunction.java:57)
	at org.renjin.primitives.special.AssignLeftFunction.apply(AssignLeftFunction.java:41)
	at org.renjin.eval.Context.evaluateCall(Context.java:407)
	at org.renjin.eval.Context.evaluate(Context.java:282)
	at org.renjin.eval.Context.evaluate(Context.java:264)
	at org.renjin.eval.Context.evaluateExpressionVector(Context.java:390)
	at org.renjin.eval.Context.evaluate(Context.java:280)
	at org.renjin.eval.Context.evaluate(Context.java:264)
	at org.renjin.repl.JlineRepl.readExpression(JlineRepl.java:244)
	at org.renjin.repl.JlineRepl.run(JlineRepl.java:131)
	at org.renjin.cli.Main.startInteractive(Main.java:126)
	at org.renjin.cli.Main.run(Main.java:111)
	at org.renjin.cli.Main.main(Main.java:78)
Caused by: java.util.UnknownFormatConversionException: Conversion = '*'
	at java.util.Formatter.checkText(Formatter.java:2579)
	at java.util.Formatter.parse(Formatter.java:2555)
	at java.util.Formatter.format(Formatter.java:2501)
	at java.util.Formatter.format(Formatter.java:2455)
	at java.lang.String.format(String.java:2940)
	at org.renjin.primitives.R$primitive$$percent$times$percent.doApply(R$primitive$$percent$times$percent.java:73)
	at org.renjin.primitives.R$primitive$$percent$times$percent.apply(R$primitive$$percent$times$percent.java:31)
	at org.renjin.eval.Context.evaluateCall(Context.java:407)
	... 83 more
```

**Comment by akbertram** [comment]
This exception is actually thrown while Renjin is trying to throw another exception... The function name "%*%" is included directly in the format argument to `sprintf()` which triggers an Exception because `%*` being interpreted as an (invalid) format specifier.

When I fix this, the underlying error message is:

```
> ypred = predict(fit, x)
ERROR: error in evaluating the argument 'x' in selecting a method for function 'as.matrix'
org.renjin.eval.EvalException: Invalid argument:
	%*%(double, S4)
	Expected:
	%*%(AtomicVector, AtomicVector)
```

It looks like the code is expecting the `%*%` operator to dispatch based on the S4 class of its second argument. The documentation from GNU R reveals this to be a new fun edge case:

> This operator is S4 generic but not S3 generic.  S4 methods need
> to be written for a function of two arguments named ‘x’ and ‘y’.

Normally, we could just mark this method as `@Generic`, but then it would also dispatch on S3 class, which it is apparently not supposed to do. I think we can update the code generator to take this distinction into account.

**Comment by akbertram** [comment]
@psolaimani I suspect this might still be missing from our S4 implementation... the ops group of functions exceptionally dispatch on both the first _and_ second argument.

**Comment by akbertram** [comment]
Fixed in  0.8.2516

**Comment by chuanwen** [comment]
Thanks for the fix!
